File: Code/Components/DanpheEMR.ServerModel/FixedAssetModels/FixedAssetStockModel.cs
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace DanpheEMR.ServerModel
{
    public class FixedAssetStockModel
    {
        [Key]
        public int FixedAssetStockId { get; set; }
        public int? GoodsReceiptItemId { get; set; }
        public virtual ItemMasterModel ItemMasterModel { get; set; }
        public int ItemId { get; set; }
        public string AssetCode { get; set; }
        public string BarCodeNumber { get; set; }
        public string AssetsLocation { get; set; }
        public string BatchNo { get; set; }
        public DateTime? WarrantyExpiryDate { get; set; }
        public decimal ItemRate { get; set; }
        public decimal? MRP { get; set; }
        public double? DiscountPercent { get; set; }
        public decimal? DiscountAmount { get; set; }
        public decimal? VATAmount { get; set; }
        public double? VAT { get; set; }
        public decimal? CcCharge { get; set; }
        public decimal? CcAmount { get; set; }
        public decimal? OtherCharge { get; set; }
        public int? CreatedBy { get; set; }
        public DateTime? CreatedOn { get; set; }
        public int? ModifiedBy { get; set; }
        public DateTime? ModifiedOn { get; set; }
        public int? CancelledBy { get; set; }
        public DateTime? CancelledOn { get; set; }
        public int? CounterId { get; set; }
        public bool? IsBarCodeGenerated { get; set; }
        public bool? IsActive { get; set; }
        public bool? IsAssetDamaged { get; set; }
        public string DamagedRemarks { get; set; }
        public string UndamagedRemarks { get; set; }
        public string Performance { get; set; }
        public int? TotalLife { get; set; }
        public double? YearOfUse { get; set; }
        public DateTime? ManufactureDate { get; set; }
        public bool IsUnderMaintenance { get; set; }
        public string SerialNo { get; set; }
        public string ModelNo { get; set; }
        public string BuildingBlockNumber { get; set; }
        public string Floors { get; set; }
        public string RoomNumber { get; set; }
        public string RoomPosition { get; set; }
        public int? DonationId { get; set; }

        public int? AssetHolderId { get; set; }
        public int StoreId { get; set; }

        public int? SubStoreId { get; set; }


        [NotMapped]
        public string CompanyPosition { get; set; }
        [NotMapped]
        public  string Name { get; set;}

        [NotMapped]
        public string PhoneNumber { get; set; }

        [NotMapped]
        public int? VendorId { get; set; }

        [NotMapped]
        public string CompanyPosition2 { get; set; }

        [NotMapped]
        public string Name2 { get; set; }

        [NotMapped]
        public string PhoneNumber2 { get; set; }

        [NotMapped]
        public string Location { get; set; }

        public bool IsMaintenanceRequired { get; set; }
        public int? ExpectedValueAfterUsefulLife { get; set; }
        public bool IsAssetDamageConfirmed { get; set; }
        public bool IsAssetScraped { get; set; }
        public decimal? ScrapAmount { get; set; }
        public string ScrapRemarks { get; set; }
        public string ScrapCancelRemarks { get; set; } 
        public int? PeriodicServiceDays { get; set; } 
        public DateTime? InstallationDate { get; set; }
        public string CssdStatus { get; set; }
        public string StockSpecification { get; set; }
    }
}

File: Code/Websites/DanpheEMR/wwwroot/DanpheApp/src/app/fixed-asset/shared/fixed-asset-stock.model.ts
import { FormGroup, FormBuilder, Validators } from "@angular/forms";
import { FLOAT } from "html2canvas/dist/types/css/property-descriptors/float";
import * as moment from "moment";

export class FixedAssetStockModel {
    public FixedAssetStockId: number = 0;
    public GoodsReceiptItemId: number = 0;
    public ItemId: number = 0;
    public AssetCode: string;
    public BarCodeNumber: string;
    public AssetsLocation: string | any;
    public BatchNo: string;
    public WarrantyExpiryDate: string = moment().format("YYYY-MM-DD");
    public IsBarCodeGenerated: boolean = false;
    public IsActive: boolean = true;
    public IsAssetDamaged: boolean = false;
    public DamagedRemarks: string;
    public UndamagedRemarks: string;
    public CreatedBy: number;
    public CreatedOn: string;

    public ItemName: string;
    public ItemCode: string;
    public VendorName: string;
    public CreatedByName: string;
    public SN: number;
    public TotalLife: number;
    public YearOfUse: FLOAT;
    public ManufactureDate: string = moment().format("YYYY-MM-DD");
    public Performance: string = "";
    public ExpectedValueAfterUsefulLife: number;

    public IsUnderMaintenance: boolean = false;
    public IsMaintenanceRequired: boolean = false;
    public IsAssetDamageConfirmed: boolean = false;
    public IsAssetScraped: boolean = false;

    public ScrapAmount: number;
    public ScrapRemarks: string;
    public ScrapCancelRemarks: string;


    public BuildingBlockNumber: string;
    public Floors: string;
    public RoomNumber: string;
    public RoomPosition: string;
    // public vendor: VendorsModel;

    public CompanyPosition: string;
    public Name: string;
    public PhoneNumber: string;
    public CompanyPosition2: string;
    public Name2: string;
    public PhoneNumber2: string;
    public VendorId: number = 0;

    public SerialNo: string;
    public ModelNo: string;
    public InstallationDate: string = moment().format("YYYY-MM-DD");
    public PeriodicServiceDays: number;
    public ServiceDate: string;
    public ServiceCompleteDate: string;

    public FAStockValidators: FormGroup = null;
    public FAEditAssetValidators: FormGroup = null;
    public FAEditAssetMainteanceValidators: FormGroup = null;

    //for Donations purpose tilganga hospital in INV_TXN_FixedAssetStock table 
    public DonationId: any;
    public Donation: string;

    //for Asset Movement Purpose
    public StoreId: number;
    public AssetHolderId: number;
    public SubStoreId: number;
    public CssdStatus: string;
    public IsColdStorageApplicable: boolean;
  IsCurrentUserMaintenanceOwner: boolean;
    constructor() {
        var _formBuilder = new FormBuilder();
        this.FAStockValidators = _formBuilder.group({
            'DamagedRemarks': ['', Validators.compose([Validators.required, Validators.maxLength(500)])],
            'ScrapAmount': ['', Validators.compose([Validators.required])],
            'ScrapRemarks': ['', Validators.compose([Validators.required, Validators.maxLength(2000)])],

        });
        this.FAEditAssetValidators = _formBuilder.group({
            'BuildingBlockNumber': ['', Validators.compose([Validators.required, Validators.maxLength(500)])],
            'SerialNo': ['', Validators.compose([Validators.required, Validators.maxLength(500)])],
            'ModelNo': ['', Validators.compose([Validators.required, Validators.maxLength(500)])],
            'AssetsLocation': ['', Validators.compose([Validators.required, Validators.maxLength(500)])],
            'Floors': ['', Validators.compose([Validators.required, Validators.maxLength(500)])],
            'RoomNumber': ['', Validators.compose([Validators.required, Validators.maxLength(500)])],
            'RoomPosition': ['', Validators.compose([Validators.required, Validators.maxLength(500)])],
            'WarrantyExpiryDate': ['', Validators.compose([])],
            'StoreId': ['', Validators.required],
            'AssetHolderId': ['']
        });

        this.FAEditAssetMainteanceValidators = _formBuilder.group({
            'CompanyPosition': ['', Validators.compose([Validators.required, Validators.maxLength(500)])],
            'Name': ['', Validators.compose([Validators.required, Validators.maxLength(500)])],
            'PhoneNumber': ['', Validators.compose([Validators.required, Validators.maxLength(500)])],
            'CompanyPosition2': ['', Validators.compose([Validators.required, Validators.maxLength(500)])],
            'Name2': ['', Validators.compose([Validators.required, Validators.maxLength(500)])],
            'PhoneNumber2': ['', Validators.compose([Validators.required, Validators.maxLength(500)])],
            'TotalLife': ['', Validators.compose([Validators.required, Validators.maxLength(500)])],
            'YearOfUse': ['', Validators.compose([Validators.required, Validators.maxLength(500)])],
            'InstallationDate': [''],
            'ManufactureDate': [''],

        });
    }

    public IsDirty(fieldName): boolean {
        if (fieldName == undefined)
            return this.FAStockValidators.dirty;
        else
            return this.FAStockValidators.controls[fieldName].dirty;
    }

    public IsValid(): boolean {
        if (this.FAStockValidators.valid) { return true; } else { return false; }
    }
    public IsValidCheck(fieldName, validator): boolean {
        if (fieldName == undefined) {
            return this.FAStockValidators.valid;
        }
        else {
            return !(this.FAStockValidators.hasError(validator, fieldName));
        }
    }


    public IsEditAssetmaintenanceDirty(fieldName): boolean {
        if (fieldName == undefined)
            return this.FAEditAssetMainteanceValidators.dirty;
        else
            return this.FAEditAssetMainteanceValidators.controls[fieldName].dirty;
    }

    public IsEditAssetmaintenanceValid(): boolean {
        if (this.FAEditAssetMainteanceValidators.valid) { return true; } else { return false; }
    }
    public IsEditAssetmaintenanceValidCheck(fieldName, validator): boolean {
        if (fieldName == undefined) {
            return this.FAEditAssetMainteanceValidators.valid;
        }
        else {
            return !(this.FAEditAssetMainteanceValidators.hasError(validator, fieldName));
        }
    }


    public IsEditAssetDirty(fieldName): boolean {
        if (fieldName == undefined)
            return this.FAEditAssetValidators.dirty;
        else
            return this.FAEditAssetValidators.controls[fieldName].dirty;
    }

    public IsEditAssetValid(): boolean {
        if (this.FAEditAssetValidators.valid) { return true; } else { return false; }
    }
    public IsEditAssetValidCheck(fieldName, validator): boolean {
        if (fieldName == undefined) {
            return this.FAEditAssetValidators.valid;
        }
        else {
            return !(this.FAEditAssetValidators.hasError(validator, fieldName));
        }
    }


}
